### PR TITLE
En 8306/stake nodes from waiting

### PR DESCRIPTION
**Pull request by sasurobert**
Move nodes from queue to staked when max number of nodes is changed

### CODE REVIEW

**Review by BeniaminDrasovean — APPROVED** [review]

File: epochStart/metachain/systemSCs.go [hunk 1]
```diff
@@ -764,6 +778,78 @@ func (s *systemSCProcessor) initDelegationSystemSC() error {
 	return nil
 }
 
+func (s *systemSCProcessor) stakeNodesFromWaitingList(nodesToStake uint32, nonce uint64) error {
```
Comment (iulianpascalau) [inline epochStart/metachain/systemSCs.go:781]: missing unit/semi-integration test

File: vm/systemSmartContracts/staking.go [hunk 2]
```diff
@@ -1406,13 +1413,134 @@ func (r *stakingSC) getOwner(args *vmcommon.ContractCallInput) vmcommon.ReturnCo
 	return vmcommon.Ok
 }
 
+func (r *stakingSC) stakeNodesFromWaitingList(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/staking.go:1416]: missing unit tests

File: vm/systemSmartContracts/staking.go [hunk 3]
```diff
@@ -1406,13 +1413,134 @@ func (r *stakingSC) getOwner(args *vmcommon.ContractCallInput) vmcommon.ReturnCo
...
+	if !r.flagStakingV2.IsSet() {
+		r.eei.AddReturnMessage("invalid method to call")
+		return vmcommon.UserError
+	}
+	if !bytes.Equal(args.CallerAddr, r.endOfEpochAccessAddr) {
+		r.eei.AddReturnMessage("this is only a view function")
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/staking.go:1422]: this is not a view function

Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:1422]: done

File: vm/systemSmartContracts/staking.go (hunk 2 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:1416]: done

File: epochStart/metachain/systemSCs.go (hunk 1 above)
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:781]: done

**Review by iulianpascalau — APPROVED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]

File: epochStart/metachain/systemSCs.go [hunk 4]
```diff
@@ -780,7 +866,7 @@ func (s *systemSCProcessor) EpochConfirmed(epoch uint32) {
 	for _, maxNodesConfig := range s.maxNodesEnableConfig {
 		if epoch >= maxNodesConfig.EpochEnable {
 			// to cover also rollbacks, we always set the maxNodes and set the Enabled flag
-			s.flagChangeMaxNodesEnabled.Toggle(true)
+			s.flagChangeMaxNodesEnabled.Toggle(epoch == maxNodesConfig.EpochEnable)
 			s.maxNodes = maxNodesConfig.MaxNumNodes
```
Comment (SebastianMarian) [inline epochStart/metachain/systemSCs.go:870]: s.maxNodes will be set to maxNodesConfig.MaxNumNodes even if epoch > maxNodesConfig.EpochEnable but 
s.flagChangeMaxNodesEnabled will be set to true only in epoch =  maxNodesConfig.EpochEnable. It is ok? If not you can revert the line 869 and change the condition in line 867 to epoch == maxNodesConfig.EpochEnable

File: vm/systemSmartContracts/staking.go [hunk 5]
```diff
@@ -60,6 +60,13 @@ type ArgsNewStakingSmartContract struct {
 	EpochNotifier        vm.EpochNotifier
 }
 
+type waitingListReturnData struct {
+	blsKeys         [][]byte
+	stakedDataList  []*StakedDataV2_0
+	lastKey         []byte
+	afterLastjailed bool
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/staking.go:67]: afterLastJailed

File: vm/systemSmartContracts/staking.go [hunk 6]
```diff
@@ -499,15 +508,26 @@ func (r *stakingSC) processStake(blsKey []byte, registrationData *StakedDataV2_0
...
+		return err
+	}
+	r.addToStakedNodes(1)
+	r.activeStakingFor(registrationData)
 
 	return nil
 }
 
+func (r *stakingSC) activeStakingFor(stakingData *StakedDataV2_0) {
+	stakingData.RegisterNonce = r.eei.BlockChainHook().CurrentNonce()
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/staking.go:523]: If you set it here now you can move line 498 after line 505 ?

File: vm/systemSmartContracts/staking.go [hunk 7]
```diff
@@ -499,15 +508,26 @@ func (r *stakingSC) processStake(blsKey []byte, registrationData *StakedDataV2_0
...
-	registrationData.StakedNonce = r.eei.BlockChainHook().CurrentNonce()
-	registrationData.UnStakedEpoch = core.DefaultUnstakedEpoch
-	registrationData.UnStakedNonce = 0
+	err := r.removeFromWaitingList(blsKey)
+	if err != nil {
+		r.eei.AddReturnMessage("error while removing from waiting")
+		return err
+	}
+	r.addToStakedNodes(1)
+	r.activeStakingFor(registrationData)
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/staking.go:517]: Do you need to r.eei.Finish([]byte{waiting/staking}) before return here?

File: vm/systemSmartContracts/staking.go [hunk 8]
```diff
@@ -1406,13 +1413,134 @@ func (r *stakingSC) getOwner(args *vmcommon.ContractCallInput) vmcommon.ReturnCo
...
+
+		blsKeysToStake = append(blsKeysToStake, element.BLSPublicKey)
+		stakedDataList = append(stakedDataList, stakedData)
+		index++
+		copy(nextKey, element.NextKey)
+	}
+
+	waitingListData.blsKeys = blsKeysToStake
+	waitingListData.stakedDataList = stakedDataList
+	waitingListData.lastKey = nextKey
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/staking.go:1533]: So, actually lastKey will remain the first element in waiting list ?

File: epochStart/metachain/systemSCs.go (hunk 4 above)
Comment (sasurobert) [reply epochStart/metachain/systemSCs.go:870]: It is ok. Only at activation epoch should the function be called and max nodes to be set

File: vm/systemSmartContracts/staking.go (hunk 7 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:517]: Return finish must be added for waiting key only.

File: vm/systemSmartContracts/staking.go (hunk 6 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:523]: No. Registration NONCE is needed when getting into the waiting list as well

Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:523]: No you can't. As you need the registration NONCE even when it got into the waiting list.

File: vm/systemSmartContracts/staking.go (hunk 8 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:1533]: Yes. All the rest is deleted. As all before when staked. Moved from waiting list to staking

**Pull request merged**